### ISSUE TITLE
Treemacs workspaces are cleared every update

### ISSUE THREAD

**Issue by benbot** [syl20bnr/spacemacs#12141]
#### Description :octocat:
Treemacs workspaces are cleared every update

#### Reproduction guide :beetle:
- Start Emacs
- Click Update Packages after treemacs publishes an update

*Observed behaviour:* :eyes: :broken_heart:
All treemacs workspaces are gone

*Expected behaviour:* :heart: :smile:
Treemacs workspaces are not gone

#### System Info :computer:
- OS: darwin
- Emacs: 27.0.50
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. 6f10e28cc)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(osx helm dash auto-completion version-control evil-snipe git github org restclient syntax-checking multiple-cursors lsp ipython-notebook notmuch
     (treemacs :variables treemacs-use-filewatch-mode t treemacs-use-follow-mode t)
     (ranger :variables ranger-override-dired t)
     (colors :variables colors-enable-nyan-cat-progress-bar t colors-colorize-identifiers 'all)
     (shell :variables shell-default-height 40 shell-default-position 'bottom shell-default-shell 'eshell)
     yaml html markdown
     (javascript :variables javascript-backend 'lsp)
     (sql :variables sql-capitalize-keywords t)
     (go :variables gofmt-command "goimports")
     emacs-lisp clojure rust elixir python ruby shell-scripts react phoenix)
```
- System configuration features: NOTIFY KQUEUE ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS NS THREADS GMP

**Comment by duianto** [comment]
Let's summon the Treemacs author @Alexander-Miller

**Comment by Alexander-Miller** [comment]
How exactly do you update treemacs?

Do you restart emacs, or do it in the current session?

Can you reproduce it by just reinstalling the package?

Is the `.emacs.d/.cache/treemacs-persist` file correctly kept between updates?

**Comment by duianto** [comment]
I can test it now, there's a Treemacs update available.

Version:
Before: treemacs-20190405.1448
After: treemacs-20190409.503

I had 4 projects in the Treemacs window:
p1
p2
p3
p4

After updating the Spacemacs packages by "clicking" (pressed `RET` with cursor) on [Update Packages] on the Spacemacs home buffer, and answering yes to update the packages.

I opened Treemacs one last time to check if the projects still were there, and they were still there (I didn't expect anything to change at this point, just wanted to make sure).

When I pressed `SPC q r` to restart, then a `*Process List*` buffer opened, and it showed:

![emacs_2019-04-09_17-20-50](https://user-images.githubusercontent.com/13420573/55812789-e5c1ab80-5aeb-11e9-8552-6d48c23c23d2.png)

Maybe the process Future is something from this package?: https://github.com/Alexander-Miller/pfuture

I clicked `y`, and Spacemacs restarted and installed the packages.

After the packages were done installing, I open Treemacs `SPC f t`, and nothing had changed, the same projects are still there:
p1
p2
p3
p4

I tried restarting `SPC q r`, and Spacemacs loaded the home buffer, and the minibuffer showed 
`[yas] Prepared just-in-time loading of snippets successfully.`, and Emacs stopped responding. The task manager showed that Emacs was using ~30% cpu.

After ~5 minutes, I force closed Emacs.

When it started again, then the Treemacs window still shows all 4 projects.

I even tried restarting a couple of more times to see if Emacs crashed again, but no, it started fine.

And the 4 projects are still shown in the Treemacs window.

#### System Info :computer:
- OS: windows-nt
- Emacs: 26.1
- Spacemacs: 0.300.0
- Spacemacs branch: develop (rev. f85acca55)
- Graphic display: t
- Distribution: spacemacs
- Editing style: vim
- Completion: helm
- Layers:
```elisp
(autohotkey auto-completion emacs-lisp git helm multiple-cursors org shell treemacs version-control)
```
- System configuration features: XPM JPEG TIFF GIF PNG RSVG SOUND NOTIFY ACL GNUTLS LIBXML2 ZLIB TOOLKIT_SCROLL_BARS THREADS LCMS2

**Comment by Alexander-Miller** [comment]
> Maybe the process Future is something from this package?

Yes, but the process futures treemacs starts should be short-lived, all they are
used for is fetching information for git highlighting and directory flattening
in a parallel python process. Even deferred git-mode adds only a 500ms delay
until the process is harvested. It's offtopic here, but I would still like to
now more about this, assuming this isn't just some non-reproducible blooper. If
you can reproduce it consistently try disabling either git-mode or setting
`treemacs-collapse-dirs` to 0. That will at least tell us which direction to look.

> and Emacs stopped responding. The task manager showed that Emacs was using ~30% cpu.

Very strange, but given that (I assume) you had yet to load treemacs something
else is happening. I would recommend you try SIGUSR2, but I don't think you can
do that in windows.

> And the 4 projects are still shown in the Treemacs window.

So there must be something more specific to the OP going on here.

@benbot I hope you can reproduce this thing somehow, because we need to dig into
this. I suppose you can make a copy of your current treemacs installation and
just move it back after every upgrade.

First we need to find out whether the persist file gets deleted or if the
loading process itself is broken. So you need to set up some projects, do an
update, and keep an eye on the persist file every step of the way, checking
whether it gets deleted or its content reset. If you find something let me know
*exactly* when that happens and what steps you took.

If that's not it then below you'll find the restore function with more logging
built in. You'll have to replace the installed version in
treemacs-persistence.el with this one and then recompile the file, *before* you
load treemacs. Then we'll see where things get stuck.

And to get the obvious out of the way: make sure that
`treemacs--should-not-run-persistence?` returns nil.

```Lisp
(defun treemacs--restore ()
  "Restore treemacs' state from `treemacs-persist-file'."
  (treemacs-log "Starting Restore")
  (unless (treemacs--should-not-run-persistence?)
    (treemacs-log "Running Restore")
    (-when-let (lines (treemacs--read-persist-lines))
      (treemacs-log "Persist Lines %s" lines)
      ;; Don't persist during restore. Otherwise, if the user would quit
      ;; Emacs during restore, for example during the completing read for
      ;; missing project action, the whole persist file would be emptied.
      (let ((kill-emacs-hook (remq #'treemacs--persist kill-emacs-hook)))
        (condition-case e
            (pcase (treemacs--validate-persist-lines lines)
              ('success
               (treemacs-log "Validation Success")
               (setf treemacs--workspaces (treemacs--read-workspaces (make-treemacs-iter :list lines))
                     (treemacs-current-workspace) (car treemacs--workspaces))
               (treemacs-log "Workspace Set To %s" (treemacs-current-workspace)))
              (`(error ,line ,error-msg)
               (treemacs--write-error-persist-state lines (format "'%s' in line '%s'" error-msg line))
               (treemacs-log "Could not restore saved state, %s:\n%s\n%s"
                             (pcase line
                               (:start "found error in the first line")
                               (:end "found error in the last line")
                               (other (format "found error in line '%s'" other)))
                             error-msg
                             (format "Broken state was saved to %s"
                                     (propertize treemacs--last-error-persist-file 'face 'font-lock-string-face)))))
          (error
           (progn
             (treemacs--write-error-persist-state lines e)
             (treemacs-log "Error '%s' when loading the persisted workspace.\n%s"
                           e
                           (format "Broken state was saved to %s"
                                   (propertize treemacs--last-error-persist-file 'face 'font-lock-string-face))))))))))
```

**Comment by duianto** [comment]
The crash isn't related to Treemacs there's an issue with the latest version of `hl-todo`.

hl-todo issue, files open modified, auto-complete layer causes hangs #12143

**Comment by benbot** [comment]
Just pulled spacemacs from develop AND updated my packages and the treemacs config was cleared out. Is there something I can do before the next time I update that would give me some logs that would help find out why this is happening?

**Comment by Alexander-Miller** [comment]
See my previous comment. You should also periodically check the contents of the persist file to see if/when it does get emptied/deleted. We need to know whether it happens when you update, when you restart emacs after the update, or when you first load a new version of treemacs.

**Comment by Alexander-Miller** [comment]
Have a look at https://github.com/Alexander-Miller/treemacs/issues/431, maybe it's the same cause.

**Comment by enikar** [comment]
Hi, good news ! The last update of treemacs (today) no longer cleared my treemacs-persist. I hope this will be true for other people. Thanks for your very good job.

**Comment by enikar** [comment]
Treemacs update is breaking again. The two last update cleared all projects in treemacs-persist.
I wait for the next update to determine if projects are cleared during the update or when I open the treemacs
windows.

**Comment by yssource** [comment]
the same problem happened to me for each time, after updating package.
with the latest develop branch of spacemacs
- M-x list-package
- `Shift-U` upgrading packages

As a result, `~/.emacs.d/.cache/treemacs-persist` is cleared every update.

**Comment by Alexander-Miller** [comment]
Well if it's a wrong directory structure then that error will be caught properly now, you will see something like this in your log:
```
[Treemacs] Could not restore saved state, found error in line ' - path :: /home/am/Documents/git/dotfiles':
Path '/home/am/Documents/git/dotfiles' appears in the workspace more than once.
Broken state was saved to /home/am/.emacs.d/.cache/treemacs-persist-at-last-error
```

**Comment by enikar** [comment]
Today, there was a treemacs update. So I tracked the deletion of projects in `treemacs-persist`.
It was not deleted while updating, but when I restarted spacemacs.

**Comment by Alexander-Miller** [comment]
So what did the log say?  What did the file actually look like before it got cleared?

**Comment by enikar** [comment]
Curiously, the last error date did'nt match the last update.
The error matched the previous update.
Anyway, this is the content of the file treemacs-persist-at-last-error:
```
# State when last error occurred on 2019-07-21 03:14:08
# Error was 'Cannot end with a project or workspace name' in line '* Default Workspace'

* Default Workspace
```
Last update was at  2019-07-22 11:56:00

**Comment by duianto** [comment]
I haven't updated yet, Is there something I could test before and after?

**Comment by Alexander-Miller** [comment]
> Error was 'Cannot end with a project or workspace name' in line '* Default Workspace'

That means that the file is empty except for this one line. We just need to figure out why.

> I haven't updated yet, Is there something I could test before and after?

Pay close attention to the content of the persist file and note when exactly its content changes - during the update, the restart, or when treemacs boots again.

BTW you should not have to wait for an actual update to try this. Just uninstalling treemacs will see spacemacs auto-install it on the next start (isn't that how spacemacs' update function works anyway?). And if that does not work the problem is likely lies in spacemacs' update routine (which is doubtful, I already looked at it a while ago and found nothing suspicious).

**Comment by enikar** [comment]
I made some experience since yesterday. I removed treemacs from
`paradox-list-packages`.
So I was able to determine when the `treemacs-persist` was cleared.
After removed treemacs package, I restarted spacemacs. So it was
reinstalled. After that, I tried to run the treemacs command.
The Projets was not cleared. Great !
But When after reinstalled treemacs packages, I chose to restart spacemacs
without runing treemacs command, then projects was cleared from the file
`treemacs-persist`.

It seems that when treemacs is reinstalled, it thinks it is the
first installation. And when spacemacs is restarted, treemacs initialization
clears the projects.

**Comment by duianto** [comment]
Confirmed.
Thanks enikar for the suggestion not to open treemacs before restarting again.

- Add a project, ex: `.emacs.d`
- Remove the `treemacs-*` directory from the elpa dir:
`C:\Users\username\.emacs.d\elpa\26.2\develop\treemacs-20190722.524`
- Open `treemacs-persist` in another editor that updates on change, for example: vscode
- Restart Spacemacs `SPC q r`, treemacs is installed. Don't open Treemacs.
- Restart `SPC q r` again, and notice that just before Spacemacs closes, then the `treemacs-persist` file changes and it now only contains:
```
* Default Workspace
```

Reproduced it in both Ubuntu and Windows.

**Comment by Alexander-Miller** [comment]
Got it. Apparently just installing treemacs does load the package, but since you never actually start the it only the default empty workspace is saved on quit. I'll make fixing this annoyance a priority.

**Comment by Alexander-Miller** [comment]
Should be fixed now.

**Comment by duianto** [comment]
Confirmed, `treemacs-persist` remains unchanged after installing Treemacs and restarting without opening Treemacs first.
